src/prof/src/lib.rs

```diff
@@ -35,6 +35,181 @@ pub struct StackProfile {
+                let program_headers =
+                    unsafe { std::slice::from_raw_parts(info.dlpi_phdr, info.dlpi_phnum.into()) };
+                let mut found_build_id = None;
+                'outer: for ph in program_headers {
+                    if ph.p_type == PT_NOTE {
+                        // From `man elf`:
+                        // typedef struct {
+                        //   Elf64_Word n_namesz;
+                        //   Elf64_Word n_descsz;
+                        //   Elf64_Word n_type;
+                        // } Elf64_Nhdr;
+                        #[repr(C)]
+                        struct NoteHeader {
+                            n_namesz: Elf64_Word,
+                            n_descsz: Elf64_Word,
+                            n_type: Elf64_Word,
+                        }
+                        // This is how `man dl_iterate_phdr` says to find the segment headers in memory.
+                        let mut offset = usize::cast_from(ph.p_vaddr + info.dlpi_addr);
+                        let orig_offset = offset;
+
+                        while offset + std::mem::size_of::<NoteHeader>() + 4
+                            <= orig_offset + usize::cast_from(ph.p_memsz)
+                        {
+                            let nh = unsafe { (offset as *const NoteHeader).as_ref() }
+                                .expect("the program headers must be well-formed");
+                            // from elf.h
+                            const NT_GNU_BUILD_ID: Elf64_Word = 3;
+                            if nh.n_type == NT_GNU_BUILD_ID && nh.n_descsz != 0 && nh.n_namesz == 4
+                            {
+                                let p_name =
+                                    (offset + std::mem::size_of::<NoteHeader>()) as *const [u8; 4];
+                                // SAFETY: since `n_namesz` is 4, the name is a four-byte value.
+                                let name = unsafe { p_name.as_ref() }.expect("this can't be null");
+                                if name == b"GNU\0" {
+                                    // We found what we're looking for!
+                                    let p_desc = (p_name as usize + 4) as *const u8;
+                                    // SAFETY: This is the documented meaning of `n_descsz`.
+                                    let desc = unsafe {
+                                        std::slice::from_raw_parts(
+                                            p_desc,
+                                            usize::cast_from(nh.n_descsz),
+                                        )
+                                    };
+                                    found_build_id = Some(desc.to_vec());
+                                    break 'outer;
+                                }
+                            }
+                            const fn align_up<const N: usize>(p: usize) -> usize {
```

Comment:
Just `(p + (N-1)) & !(N-1)` might be more idiomatic?

Reply:
probably slightly more idiomatic, but I don't like forcing readers to think about bit twiddling if not actually necessary

Reply:
heh, I only say that because it took me longer to convince myself this function was correct than what I'm used to, but that's just personal preference I guess.